File: leetcode/AIcoding/3.DeepSeek-v3/javascript_solutions/74-search-a-2d-matrix.js
var searchMatrix = function(matrix, target) {
    if (!matrix.length || !matrix[0].length) return false;
    let m = matrix.length, n = matrix[0].length;
    let left = 0, right = m * n - 1;
    while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        let midVal = matrix[Math.floor(mid / n)][mid % n];
        if (midVal === target) return true;
        else if (midVal < target) left = mid + 1;
        else right = mid - 1;
    }
    return false;
};

File: leetcode/AIcoding/3.DeepSeek-v3/rust_solutions/74-search-a-2d-matrix.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let m = matrix.len();
        if m == 0 {
            return false;
        }
        let n = matrix[0].len();
        if n == 0 {
            return false;
        }
        let mut left = 0;
        let mut right = m * n - 1;
        while left <= right {
            let mid = left + (right - left) / 2;
            let mid_val = matrix[mid / n][mid % n];
            if mid_val == target {
                return true;
            } else if mid_val < target {
                left = mid + 1;
            } else {
                if mid == 0 {
                    break;
                }
                right = mid - 1;
            }
        }
        false
    }
}